sql/core/src/main/scala/org/apache/spark/sql/internal/BaseSessionStateBuilder.scala

```diff
@@ -263,6 +266,14 @@ abstract class BaseSessionStateBuilder(
    */
   protected def customEarlyScanPushDownRules: Seq[Rule[LogicalPlan]] = Nil
 
+  /**
+   * Custom rules for rewriting plans for v2 tables to add to the Optimizer. Prefer overriding
+   * this instead of creating your own Optimizer.
+   *
+   * Note that this may NOT depend on the `optimizer` function.
+   */
+  protected def customV2SourceRewriteRules: Seq[Rule[LogicalPlan]] = Nil
```

Comment:
I am going to add a way to inject a custom rule through extensions here in a follow-up PR.